File: src/state/capabilities.ts
import type { Capabilities } from "../bindings/Capabilities";
import type { ProviderInfo } from "../bindings/ProviderInfo";
import type { SessionState } from "../bindings/SessionState";
import type { Support } from "../bindings/Support";

export type Action = "stop" | "sendPrompt" | "permissions" | "launch";

export interface ActionAvailability {
  enabled: boolean;
  /** Why the action is disabled (shown as a tooltip), or the support level badge. */
  reason: string;
  support: Support;
}

/** Capabilities for the session's mode, plus whether this build implements that mode. */
export function modeCapabilities(
  provider: ProviderInfo | undefined,
  mode: SessionState["mode"],
): { caps: Capabilities | null; implemented: boolean } {
  if (!provider) return { caps: null, implemented: false };
  const profile = provider.capabilities;
  return mode === "managed"
    ? { caps: profile.managed, implemented: profile.implemented.managedSessions }
    : { caps: profile.external, implemented: profile.implemented.externalSessions };
}

/**
 * Decides whether an action button is enabled. Actions are enabled only when
 * the provider supports the capability for this session mode AND this build
 * implements it. Nothing is ever simulated for a real provider.
 */
export function actionAvailability(
  provider: ProviderInfo | undefined,
  mode: SessionState["mode"],
  action: Action,
): ActionAvailability {
  const { caps, implemented } = modeCapabilities(provider, mode);
  if (!caps) return { enabled: false, reason: "Unknown provider", support: "unsupported" };
  const support = caps[action];
  const name = provider?.descriptor.displayName ?? "This provider";
  if (support === "unsupported") {
    return { enabled: false, reason: `${name} does not support this for ${mode} sessions`, support };
  }
  if (!implemented) {
    return { enabled: false, reason: `${name} ${mode} sessions are not implemented yet in this build`, support };
  }
  return { enabled: true, reason: support === "supported" ? "" : `Support level: ${support}`, support };
}

File: crates/ao-core/src/capabilities.rs
//! Explicit, honest capability declarations. See docs/PROVIDER_CAPABILITIES.md.

use serde::{Deserialize, Serialize};
use ts_rs::TS;

/// How well a provider supports a capability.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "lowercase")]
#[ts(export)]
pub enum Support {
    /// Official, documented and verified.
    Supported,
    /// Available with a documented limitation.
    Partial,
    /// Exists upstream but experimental or reported incomplete.
    Experimental,
    /// Negotiated per session at runtime (e.g. ACP `initialize`).
    Runtime,
    /// Not available. Never simulated.
    Unsupported,
}

impl Support {
    /// Whether the UI may offer the feature (possibly with a badge).
    pub fn is_available(self) -> bool {
        !matches!(self, Support::Unsupported)
    }
}

/// Capability set for one integration mode (managed or external sessions).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct Capabilities {
    pub launch: Support,
    pub attach: Support,
    pub list_sessions: Support,
    pub stop: Support,
    pub send_prompt: Support,
    pub structured_events: Support,
    pub tool_events: Support,
    pub file_events: Support,
    pub command_events: Support,
    pub permissions: Support,
    pub subagents: Support,
    pub usage: Support,
    pub cost: Support,
    pub model: Support,
    pub context_compaction: Support,
    pub resume: Support,
}

impl Capabilities {
    /// Everything unsupported. Adapters start from here and opt in explicitly.
    pub const NONE: Capabilities = Capabilities {
        launch: Support::Unsupported,
        attach: Support::Unsupported,
        list_sessions: Support::Unsupported,
        stop: Support::Unsupported,
        send_prompt: Support::Unsupported,
        structured_events: Support::Unsupported,
        tool_events: Support::Unsupported,
        file_events: Support::Unsupported,
        command_events: Support::Unsupported,
        permissions: Support::Unsupported,
        subagents: Support::Unsupported,
        usage: Support::Unsupported,
        cost: Support::Unsupported,
        model: Support::Unsupported,
        context_compaction: Support::Unsupported,
        resume: Support::Unsupported,
    };
}

/// Capabilities for both integration modes plus human-readable notes.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct CapabilityProfile {
    pub managed: Capabilities,
    pub external: Capabilities,
    /// Whether the adapter itself is implemented in this build. Capabilities
    /// describe what the provider offers; `implemented` says what Agent Office
    /// can already use. The UI enables actions only when both are true.
    pub implemented: ImplementedFeatures,
    pub notes: Vec<String>,
}

/// What this build of Agent Office has implemented for a provider.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct ImplementedFeatures {
    pub detection: bool,
    pub managed_sessions: bool,
    pub external_sessions: bool,
    pub integration_setup: bool,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn none_is_fully_unsupported() {
        let json = serde_json::to_value(Capabilities::NONE).unwrap();
        for (_, v) in json.as_object().unwrap() {
            assert_eq!(v, "unsupported");
        }
        assert!(!Support::Unsupported.is_available());
        assert!(Support::Experimental.is_available());
    }
}
